WAM broker update docs (#131)
MicrosoftDocs/microsoft-authentication-library-dotnet: msal-dotnet-articles/acquiring-tokens/desktop-mobile/wam.md

---
title: Using MSAL.NET with Web Account Manager (WAM)

> description: "MSAL is able to call Web Account Manager (WAM), a Windows component that ships with the OS. This component acts as an authentication broker allowing the users of your app  to benefit from integration with accounts known to Windows, such as the account you signed into your Windows session."

Reviewer: ```suggestion
description: "MSAL is able to call Web Account Manager (WAM), a Windows component that ships with the OS. This component acts as an authentication broker allowing the users of your app to benefit from integration with accounts known to Windows, such as the account you signed into your Windows session."
```

---

# Using MSAL.NET with Web Account Manager (WAM)

> MSAL is able to call Web Account Manager (WAM), a Windows component that ships with the OS. This component acts as an authentication broker allowing the users of your app  to benefit from integration with accounts known to Windows, such as the account you signed into your Windows session.

Reviewer: ```suggestion
MSAL is able to call Web Account Manager (WAM), a Windows component that ships with the OS. This component acts as an authentication broker allowing the users of your app to benefit from integration with accounts known to Windows, such as the account you signed into your Windows session.
```

>[!NOTE]
>WAM is only available on Windows 10 and above, as well as Windows Server 2019 and above.

## What is a broker

> An authentication broker is an application that runs on a user’s machine that manages the authentication handshakes and token maintenance for connected accounts. The windows operating system uses the Web Account Manager (WAM) as its authentication broker. It has many benefits for developers and customers alike, including:

Reviewer: ```suggestion
An authentication broker is an application that runs on a user’s machine that manages the authentication handshakes and token maintenance for connected accounts. The Windows operating system uses the Web Account Manager (WAM) as its authentication broker. It has many benefits for developers and customers alike, including:
```

With the help of a broker you need to write less code to handle token-related logic while also being able to use more advanced functionality, such as [Proof-of-Possession tokens](../../advanced/proof-of-possession-tokens.md). Moving forward, the MSAL team is continuing to invest in making sure that brokers are the de facto approach to authenticate inside applications.

## Improved Windows broker experience

Latest Windows releases include an updated WAM. The new broker is written in C++, is well-tested, and is significantly more performant and secure. One of the biggest consumers of the new broker experience is the Microsoft 365 suite of apps.

> [!IMPORTANT]
> With the introduction of the updated broker, we have also updated MSAL.NET to expose its capabilities under a simplified API. Starting with version 4.52.0, the new broker will be the default in MSAL.NET.

The new WAM fixes a number of issues with the legacy implementation and provides other benefits, including:

- New implementation is more stable, easier to add new features, and has less chance of regressions.

> - Works in apps that are ran under the Administrator user context.

Reviewer: ```suggestion
- Works in apps that are run under the Administrator user context.
```

Author: I am actually curious about the right grammar here, @gladjohn 😄 Maybe the right word here is "executed."

- Adds support for [Proof-of-Possession tokens](../../advanced/proof-of-possession-tokens.md).
- Decreases assembly size.

## Enabling WAM

Due to platform-specific and backwards compatibility requirements, WAM support is split across three packages:

- [Microsoft.Identity.Client](https://www.nuget.org/packages/Microsoft.Identity.Client/) (i.e., MSAL) - core library for token acquisition.
- [Microsoft.Identity.Client.Broker](https://www.nuget.org/packages/Microsoft.Identity.Client.Broker/) - adds support for authentication with the broker.

> - [Microsoft.Identity.Client.Desktop](https://www.nuget.org/packages/Microsoft.Identity.Client.Desktop/) - adds support for authentication with the broker on legacy platforms, like .NET Framework, .NET Core 3.1, and UWP.

Reviewer: we should be able to use new broker in .NET FWK and .Net Core 3.1

Author: @gladjohn - would that be through `Microsoft.Identity.Client.Desktop` or are you saying that we can natively do it through the broker package? If that is the case, what is the purpose of the desktop package?

After referencing the relevant packages, call [`WithBroker(BrokerOptions)`](xref:Microsoft.Identity.Client.Desktop.WamExtension.WithBroker*) with broker configuration options and [a window handle](#parent-window-handles) that the broker will be bound to.

```csharp
BrokerOptions options = new BrokerOptions(BrokerOptions.OperatingSystems.Windows);

options.Title = "My Awesome Application";
options.ListOperatingSystemAccounts = true;

> IPublicClientApplication app =

Reviewer: adding authority and adding a link guidance [here](https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-client-application-configuration), will help with teams adopting to use WAM

    PublicClientApplicationBuilder.Create("YOUR_CLIENT_ID")
    .WithDefaultRedirectUri()
    .WithParentActivityOrWindow(GetConsoleOrTerminalWindow)
    .WithBroker(options)
    .Build();

var authResult = await app.AcquireTokenInteractive(new List<string>() { "User.Read" }).ExecuteAsync();
```

When using the broker, if the authority used is targeting Azure AD as well as personal Microsoft accounts, the user will first be prompted to select an account using the built-in system account picker.

![Demo of the WAM component](../../media/wam/wam-demo.gif)

If the configuration is set on a per-tenant basis by using [`WithTenantId`](xref:Microsoft.Identity.Client.AbstractApplicationBuilder`1.WithTenantId*) or if the authority is set to an [audience](xref:Microsoft.Identity.Client.AadAuthorityAudience) that _does not_ include personal Microsoft accounts, the native Windows account picker will not be shown and instead the user will be prompted with a generic Microsoft authentication prompt.

![Demo of the WAM component that is configured on a per-tenant basis and doesn't show the OS-based account picker](../../media/wam/wam-per-tenant.gif)

> Once the account is added or selected, the user will be prompted for additional consent if the have never used the application before or the application requires additional permissions.

Reviewer: ```suggestion
Once the account is added or selected, the user will be prompted for additional consent if they have never used the application before or the application requires additional permissions.
```

```

When using the broker, if the [authority](/azure/active-directory/develop/msal-client-application-configuration#authority) used is targeting Azure AD as well as personal Microsoft accounts, the user will first be prompted to select an account using the built-in system account picker.

![Demo of the WAM component](../../media/wam/wam-demo.gif)

If the configuration is set on a per-tenant basis by using [`WithTenantId`](xref:Microsoft.Identity.Client.AbstractApplicationBuilder`1.WithTenantId*) or if the authority is set to an [audience](xref:Microsoft.Identity.Client.AadAuthorityAudience) that _does not_ include personal Microsoft accounts, the native Windows account picker will not be shown and instead the user will be prompted with a generic Microsoft authentication prompt.

![Demo of the WAM component that is configured on a per-tenant basis and doesn't show the OS-based account picker](../../media/wam/wam-per-tenant.gif)

Once the account is added or selected, the user will be prompted for additional consent if they have never used the application before or the application requires additional permissions.

>[!NOTE]

> >No changes are required for UWP applications. Because the platform does not support the updated broker, existing applications will continue to use the legacy WAM implementation, which is documented in [Acquire a token using WAM](/azure/active-directory/develop/scenario-desktop-acquire-token-wam).

Reviewer: ```suggestion
>No changes are required for UWP applications. Because the platform does not support the updated broker, existing applications will continue to use the legacy WAM implementation, which is documented in [Acquire a token using WAM](/azure/active-directory/develop/scenario-desktop-acquire-token-wam).
```
the link still leads to a page that talks about the new broker

Author: At some point we need to merge that page into the current one. I removed the reference for now.

## Parent window handles

> To use the broker, it is now required to provide the window handle to which the WAM modal dialog be parented using [`WithParentActivityOrWindow`](xref:Microsoft.Identity.Client.PublicClientApplicationBuilder.WithParentActivityOrWindow*) APIs. Trying to infer a window is not feasible and in the past this has led to bad user experiences where the authentication window was hidden behind the application window.

Reviewer 2: ```suggestion
To use the broker, it is now required to provide the window handle to which the WAM modal dialog be parented using [`WithParentActivityOrWindow`](xref:Microsoft.Identity.Client.PublicClientApplicationBuilder.WithParentActivityOrWindow*) APIs. The window handle must be provided by the developer because it's not feasible for MSAL itself to infer the parent window and in the past this has led to bad user experiences where the authentication window was hidden behind the application window.
```

## Proof-of-Possession access tokens

The WAM broker allows acquiring PoP tokens for public client flows. See [Proof-of-Possession tokens](../../advanced/proof-of-possession-tokens.md) for more details.

## Redirect URI

WAM redirect URIs do not need to be configured in MSAL, but they must be configured in the app registration. They should follow the pattern below:

```text
ms-appx-web://microsoft.aad.brokerplugin/{client_id}
```

>[!NOTE]
>When configuring the redirect URL in the Azure Portal, ensure that you're setting it in the **Mobile and desktop applications** section.

## Username/password flow

> This flow, also known as Resource Owner Password Credentials (ROPC), is not recommended except in test scenarios or in scenarios where service principal access to a resource gives it too much access and you can only scope it down with user flows. When using WAM, `AcquireTokenByUsernamePassword` will let WAM manage the protocol and fetch tokens.

Reviewer 2: nit: add an xref code link?

>[!WARNING]

> >There are a few important considerations that you need to account for when using the ROPC flow. One of the main ones is that it is **not supported for personal Microsoft accounts** and for **Azure AD accounts with multi-factor authentication enabled**. Check out [Microsoft identity platform and OAuth 2.0 Resource Owner Password Credentials](/azure/active-directory/develop/v2-oauth-ropc) for the full overview.

Reviewer 2: ```suggestion
>There are a few important considerations that you need to account for when using the ROPC flow. One of the main ones is that it **doesn't support personal Microsoft accounts** and **Azure AD accounts with enabled multi-factor authentication**. Check out [Microsoft identity platform and OAuth 2.0 Resource Owner Password Credentials](/azure/active-directory/develop/v2-oauth-ropc) for the full overview.
```
